<a name="ilp"></a>
## Import necessary libraries and packages 

In [1]:
import pandas as pd
import os
import numpy as np

import tensorflow as tf
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Embedding, LSTM, Dense, Bidirectional
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam

## Dataset information

In [2]:
medium_data = pd.read_csv('./medium_data.csv')
medium_data.head()

id                                                url  \
0   1  https://towardsdatascience.com/a-beginners-gui...   
1   2  https://towardsdatascience.com/hands-on-graph-...   
2   3  https://towardsdatascience.com/how-to-use-ggpl...   
3   4  https://towardsdatascience.com/databricks-how-...   
4   5  https://towardsdatascience.com/a-step-by-step-...   

                                               title  \
0  A Beginner’s Guide to Word Embedding with Gens...   
1  Hands-on Graph Neural Networks with PyTorch & ...   
2                       How to Use ggplot2 in Python   
3  Databricks: How to Save Files in CSV on Your L...   
4  A Step-by-Step Implementation of Gradient Desc...   

                                  subtitle   image  claps responses  \
0                                      NaN   1.png    850         8   
1                                      NaN   2.png   1100        11   
2         A Grammar of Graphics for Python   3.png    767         1   
3  When I work on Python projects dealing…  4.jpeg    354         0   
4          One example of building neural…  5.jpeg    211         3   

   reading_time           publication        date  
0             8  Towards Data Science  2019-05-30  
1             9  Towards Data Science  2019-05-30  
2             5  Towards Data Science  2019-05-30  
3             4  Towards Data Science  2019-05-30  
4             4  Towards Data Science  2019-05-30

In [3]:
print("Number of records: ", medium_data.shape[0])
print("Number of fields: ", medium_data.shape[1])

Number of records:  6508
Number of fields:  10


Here, we have a 10 different fields and 6508 records but we will only use title field for predicting next word. 

## Display titles of various articles  and preprocess them

In [4]:
medium_data['title']

0       A Beginner’s Guide to Word Embedding with Gens...
1       Hands-on Graph Neural Networks with PyTorch & ...
2                            How to Use ggplot2 in Python
3       Databricks: How to Save Files in CSV on Your L...
4       A Step-by-Step Implementation of Gradient Desc...
                              ...                        
6503    “We” vs “I” — How Should You Talk About Yourse...
6504                     How Donald Trump Markets Himself
6505        Content and Marketing Beyond Mass Consumption
6506    5 Questions All Copywriters Should Ask Clients...
6507               How To Write a Good Business Blog Post
Name: title, Length: 6508, dtype: object

#### Removing unwanted characters and words in titles

In [5]:
medium_data['title'] = medium_data['title'].apply(lambda x: x.replace(u'\xa0',u' '))
medium_data['title'] = medium_data['title'].apply(lambda x: x.replace('\u200a',' '))

## Tokenzation

In [6]:
tokenizer = Tokenizer(oov_token='<oov>') # For those words which are not found in word_index
tokenizer.fit_on_texts(medium_data['title'])
total_words = len(tokenizer.word_index) + 1

print("Total number of words: ", total_words)
print("Word: ID")
print("------------")
print("<oov>: ", tokenizer.word_index['<oov>'])
print("Strong: ", tokenizer.word_index['strong'])
print("And: ", tokenizer.word_index['and'])
print("Consumption: ", tokenizer.word_index['consumption'])

Total number of words:  8238
Word: ID
------------
<oov>:  1
Strong:  4
And:  8
Consumption:  8237


## Titles text into sequences and make n_gram model

In [7]:
input_sequences = []
for line in medium_data['title']:
    token_list = tokenizer.texts_to_sequences([line])[0]
    #print(token_list)
    
    for i in range(1, len(token_list)):
        n_gram_sequence = token_list[:i+1]
        input_sequences.append(n_gram_sequence)

# print(input_sequences)
print("Total input sequences: ", len(input_sequences))

Total input sequences:  48461


## Making all titles with same length by using padding

In [8]:
# pad sequences 
max_sequence_len = max([len(x) for x in input_sequences])
input_sequences = np.array(pad_sequences(input_sequences, maxlen=max_sequence_len, padding='pre'))
input_sequences[1]

array([  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
         0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   5, 676,
        68])

## Preparing features and labels

In [9]:
# create features and label
xs, labels = input_sequences[:,:-1],input_sequences[:,-1]
ys = tf.keras.utils.to_categorical(labels, num_classes=total_words)

In [10]:
print(xs[5])
print(labels[5])
print(ys[5][14])

[   0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    5  676   68    2  452 1518]
14
1.0


## Bi- LSTM Neural Network Model training

In [ ]:
model = Sequential()
model.add(Embedding(total_words, 100, input_length=max_sequence_len-1))
model.add(Bidirectional(LSTM(150)))
model.add(Dense(total_words, activation='softmax'))
adam = Adam(learning_rate=0.01)
model.compile(loss='categorical_crossentropy', optimizer=adam, metrics=['accuracy'])
history = model.fit(xs, ys, epochs=50, verbose=1)
print(model)

C:\Users\adity\OneDrive - IIT Delhi\Anaconda installation\lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Epoch 1/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 90s 57ms/step - accuracy: 0.1007 - loss: 7.0448
Epoch 2/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 88s 58ms/step - accuracy: 0.1726 - loss: 5.7145
Epoch 3/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 91s 60ms/step - accuracy: 0.2140 - loss: 4.8212
Epoch 4/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 95s 63ms/step - accuracy: 0.2653 - loss: 4.0216
Epoch 5/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 93s 61ms/step - accuracy: 0.3252 - loss: 3.4258
Epoch 6/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 94s 62ms/step - accuracy: 0.3655 - loss: 3.0765
Epoch 7/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 93s 61ms/step - accuracy: 0.3980 - loss: 2.8327
Epoch 8/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 134s 56ms/step - accuracy: 0.4308 - loss: 2.6435
Epoch 9/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 88s 58ms/step - accuracy: 0.4475 - loss: 2.4973
Epoch 10/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 90s 60ms/step - accuracy: 0.4591 - loss: 2.4205
Epoch 11/50
1515/1515 ━━━━━━━━━━━━━━━━━━━━ 83s 55ms/step - accuracy: 0.4709 - loss: 2.3600
Epoch 1

<a name="acc"></a>
# Plotting model accuracy and loss

In [ ]:
import matplotlib.pyplot as plt


def plot_graphs(history, string):
    plt.plot(history.history[string])
    plt.xlabel("Epochs")
    plt.ylabel(string)
    plt.show()

In [ ]:
plot_graphs(history, 'accuracy')

In [ ]:
plot_graphs(history, 'loss')

<a name="new"></a>
# Predicting next word of title

In [ ]:
seed_text = "implementation of"
next_words = 2
  
for _ in range(next_words):
    token_list = tokenizer.texts_to_sequences([seed_text])[0]
    token_list = pad_sequences([token_list], maxlen=max_sequence_len-1, padding='pre')
    predicted = model.predict_classes(token_list, verbose=0)
    output_word = ""
    for word, index in tokenizer.word_index.items():
        if index == predicted:
            output_word = word
            break
    seed_text += " " + output_word
print(seed_text)